### PR TITLE
Fix sorting logic for best tiles to load

### PR DESCRIPTION
**Pull request by pearcetm**
The logic in `TiledImage._sortTiles` incorrectly sorts by level visibility, which prioritizes high-resolution tiles over lower ones, leading to the issue described in https://github.com/openseadragon/openseadragon/issues/2533.  

Fixes https://github.com/openseadragon/openseadragon/issues/2533 by changing the sort order based on level visibility.

In addition, the order of the first two arguments passed to `_updateTile` is swapped compared to the definition.
https://github.com/openseadragon/openseadragon/blob/f3a942c5224012d9bfca2ba4bef71796473cb60d/src/tiledimage.js#L1653-L1655
vs
https://github.com/openseadragon/openseadragon/blob/f3a942c5224012d9bfca2ba4bef71796473cb60d/src/tiledimage.js#L1744

I'm not entirely certain what the implications of this are for performance, but fixing it seems like a good idea.

@thec0keman can you give this PR a try and see if it fixes your issue?

### CODE REVIEW

**Comment by pearcetm** [comment]
For reference, I believe the low-res tiles not being prioritized is the same thing I noticed here: https://github.com/openseadragon/openseadragon/pull/2387#issuecomment-1662663869

I didn't realize that it wasn't fixed before that PR was merged. But this one should fix it.

**Review by iangilman — APPROVED** [review]
Excellent... @pearcetm Thank you for tracking this down, and thank you to @thec0keman for reporting it!

I'm approving this but won't merge it yet until we hear from @thec0keman.

**Comment by pearcetm** [comment]
> It had some unintended consequences in our setup. While panning, old tiles are rendered in unexpected places

Oh, that's funky! I had only tested with the `webgl` drawer - which seems to work fine - but the `canvas` drawer has the issue you describe here. I'll look into why canvas is misbehaving with this change.

**Comment by pearcetm** [comment]
@antond I just reverted the change that swapped the order of arguments in the call to `_updateTile` mentioned above. On my system, this has fixed the behavior of the `canvas` drawer that you noticed and I was able to reproduce. 

Now the behavior I see is:
1) low-res tiles are prioritized for loading, with higher-res tiling in later
2) tiles are loaded from center of the viewport outwards
3) no positioning artifacts in either drawer type

If you could give this new change a try I'd appreciate it.

**Comment by pearcetm** [comment]
> I'm not entirely certain what the implications of this are for performance, but fixing it seems like a good idea.

Turns out this was **not** a good idea :). I have reverted it.

I am kind of at a loss as to what exactly is going on here - this was originally in the legacy drawing code and I don't fully understand what is happening. I feel like at the very least we need to change the names of these parameters and document better what they do. Thoughts, @iangilman?

**Comment by pearcetm** [comment]
I made the order of the arguments consistent between `_updateLevel` and `_updateTile`, and swapped the variables within `_updateTile` to keep the logic the same as it was previously when it was working. It still seems to work (which is good).

Having done this, the logic (or lack of logic ;) ) of the pipeline got a little more clear... The code snippets below are modified from current master, not this PR which is in flux.

In `_updateLevelsForViewport`:
```
_updateLevelsForViewport: function(){
        ...
        var haveDrawn = false; // we start by setting this flag to false
        ...
        // Update any level that will be drawn
        for (let i = 0; i < levelList.length; i++) {
            let level = levelList[i];
            var drawLevel = false;
            ...
            // Here, we either set haveDrawn = true, or skip the rest of the loop if it isn't true
            if (i === levelList.length - 1 || (!haveDrawn && currentRenderPixelRatio >= this.minPixelRatio) ) {
                drawLevel = true;
                haveDrawn = true;
            } else if (!haveDrawn) {
                continue; // if haveDrawn == false, skip the rest of this loop
            }

           // which means by here, haveDrawn is always true
          ...
          var result = this._updateLevel(
                haveDrawn, // <-- This is always true
                drawLevel,
                level,
                levelOpacity,
                levelVisibility,
                drawArea,
                currentTime,
                bestTiles
            );
           ...
     }
```
So, the first argument to `_updateLevel` is **always true**

```
_updateLevel: function(haveDrawn, drawLevel, level, levelOpacity, levelVisibility, drawArea, currentTime, best) {
        // haveDrawn is always true, because of the guard in _updateLevelsForViewport which calls this function
        ...
        for (var x = topLeftTile.x; x <= bottomRightTile.x; x++) {
            for (var y = topLeftTile.y; y <= bottomRightTile.y; y++) {
                ...
                var result = this._updateTile(
                    drawLevel,
                    haveDrawn, // <----- the second argument to _updateTile is always true
                    flippedX, y,
                    level,
                    levelVisibility,
                    viewportCenter,
                    numberOfTiles,
                    currentTime,
                    best
                );
                ...
            }
        }
```
So, because the arguments were swapped, the **second** argument passed into `_updateTiles` is now always true.
```
_updateTile: function( haveDrawn, drawLevel, x, y, level, levelVisibility, viewportCenter, numberOfTiles, currentTime, best){
        // haveDrawn (our first argument) takes the value of drawLevel from _updateLevelsFromViewport
        // which is true ONLY for the level during which haveDrawn is changed from false to true 

        // drawLevel (our second argument) is always true
        var drawTile = drawLevel; // drawTile just aliases drawLevel; drawTile is always true
        ...
        // The below block never gets called, because drawTile is never false
        if ( haveDrawn && !drawTile ) {
            if ( this._isCovered( this.coverage, level, x, y ) ) {
                this._setCoverage( this.coverage, level, x, y, true );
            } else {
                drawTile = true;
            }
        }
        // Same here - drawTile is never false, so this never gets called
        if ( !drawTile ) {
            return {
                bestTiles: best,
                tile: tile
            };
        }
        
        // Which means that _positionTile always gets called (so tiles are always positioned correctly)
        this._positionTile(
            tile,
            this.source.tileOverlap,
            this.viewport,
            viewportCenter,
            levelVisibility
        );
```

So, when I swapped the arguments in `_updateLevel` to match the signature of `_updateTile`, all of a sudden `_drawTile` could be false, and the tiles weren't always getting positioned... leading to the visual glitches @antond showed with the video above.

I'm going to try to clean this up a bit by removing the code that was never accessed, and dropping the `haveDrawn` arguments which are never used.

**Comment by pearcetm** [comment]
Once I got rid of `haveDrawn`, it turns out that `drawLevel` and `drawTile` weren't actually used for anything, so I was able to ditch those too. This let me unravel a bit more spaghetti code - there were more calculations and guards that were entirely unused because of other changes, so I cleaned those up too. I also added some comments to hopefully clarify a bit what is going on with the parts that are left.

File: src/tiledimage.js
Comment (iangilman) [reply src/tiledimage.js:1370]: I see. Presumably you ran some tests and observed this? Is there a chance it would not be the case in some circumstances? It doesn't seem like it hurts anything to have the guard here just in case.

File: src/tiledimage.js
Comment (iangilman) [reply src/tiledimage.js:1372]: Sounds like we agree. `haveDrawn` is always true, `drawLevel` could either be true or false.

File: src/tiledimage.js
Comment (iangilman) [reply src/tiledimage.js:1787]: Yes, good point! I was focusing on `haveDrawn` here and not paying attention to the fact `!drawTile` would never happen.

Comment (pearcetm) [reply src/tiledimage.js:1787]: > Yes, good point! I was focusing on `haveDrawn` here and not paying attention to the fact `!drawTile` would never happen.

Yeah, it is convoluted! Once I deleted this `if()` statement, that's when the argument `haveDrawn` (which comes from `drawLevel` in the calling function) became unused.

Are you in agreement that this can be dropped, and that the upstream `drawLevel` can therefore also be dropped?

File: src/tiledimage.js
Comment (pearcetm) [reply src/tiledimage.js:1370]: > I see. Presumably you ran some tests and observed this? Is there a chance it would not be the case in some circumstances? It doesn't seem like it hurts anything to have the guard here just in case.

So I did do some testing and I did observe this. But I also looked at how the calculation is done within `getLevelsInterval`. I think that division by `this.minPixelRatio` within the `Math.log`, followed by `Math.floor` which rounds down, guarantees that `currentRenderPixelRatio >= this.minPixelRatio` is true for all the levels we're iterating over - because if it weren't, the ratio inside the `Math.log` call would fall below 1, causing `Math.log` to be negative, and we'd round down to the next lower level as the highest one to start with.

So I think it is mathematically defined that this guard is never going to be triggered.

File: src/tiledimage.js
Comment (iangilman) [reply src/tiledimage.js:1787]: Yes, I agree. Thank you for confirming :)

File: src/tiledimage.js
Comment (iangilman) [reply src/tiledimage.js:1370]: I'm afraid it's too complicated for me to confidently say it would never happen. I guess I'm wondering what benefit we get from removing this clause... If it's just some redundant safety, maybe it's worth keeping in? Or are you thinking it's actually causing problems?

File: src/tiledimage.js
Comment (pearcetm) [reply src/tiledimage.js]: Done

File: src/tiledimage.js
Comment (pearcetm) [reply src/tiledimage.js:1370]: I don't think it was causing any problems with code execution (since I don't believe it was doing anything at all). The reason I wanted to remove it was to simplify the code so it is more understandable, since more understandable is more maintainable. `_getLevelsInterval` is only used in one place, which is to calculated the levels to use in this loop. If that were incorrect, I think ideally it should be fixed there so that function gave accurate levels, rather than second-guessing that calculation during the loop itself. (Note: I don't think that function is incorrect or I would have tried to fix it).

Regardless, I've added the guard back in. There are slightly different variables names and it is a little more streamlined, but it does the same thing it did before.

**Review by iangilman — APPROVED** [review]
I appreciate the cleanup, and I agree, untangling the mess is worthwhile! Thank you for indulging my paranoia in that one spot... I suppose there's a fine line between superstition and sensible caution. 

I think this is ready to merge. I think #2533 has been resolved (though this ended up being about more than just that). Would you agree?

**Comment by pearcetm** [comment]
@iangilman I agree - this fixes #2533 by changing `return (a.visibility - b.visibility);` to `return (b.visibility - a.visibility);`, which now correctly prioritizes loading big low-res tiles first before smaller high-res tiles. The rest of that issue was resolved separately, and the other changes here just clean up the code a bit.

> Thank you for indulging my paranoia in that one spot...

No problem! Leaving the check in isn't expensive, and you're right that it is tough to sort out exactly what is going on with the math behind the level calculations, so it is hard to be sure that it would never be needed.

**Comment by iangilman** [comment]
Great! Merged.

> No problem! Leaving the check in isn't expensive, and you're right that it is tough to sort out exactly what is going on with the math behind the level calculations, so it is hard to be sure that it would never be needed.

Another related thought: even if the logic is airtight right now, we have to keep in mind that other aspects of the code may be changed in the future, leading to different conditions here.

**Pull request merged**
